### PR TITLE
Add AccessHandle proposal

### PR DESCRIPTION
**Pull request by fivedots**
This PR adds the proposal document for AccessHandles. AccessHandles are a new proposed surface that allows us to support [Storage Foundation API's](https://github.com/WICG/storage-foundation-api-explainer) use cases through the Origin Private File System.

Further details on the origins of this proposal, and alternatives considered, can be found in the following documents: [Merging Storage Foundation API and the Origin Private File System](https://docs.google.com/document/d/121OZpRk7bKSF7qU3kQLqAEUVSNxqREnE98malHYwWec), [Recommendation for Augmented OPFS](https://docs.google.com/document/d/1g7ZCqZ5NdiU7oqyCpsc2iZ7rRAY1ZXO-9VoG4LfP7fM).

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/file-system-access/pull/310.html" title="Last updated on Jul 27, 2021, 5:51 PM UTC (e88a265)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/file-system-access/310/3309115...e88a265.html" title="Last updated on Jul 27, 2021, 5:51 PM UTC (e88a265)">Diff</a>

### CODE REVIEW

File: AccessHandle.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+  - [AccessHandle IDL](#accesshandle-idl)
+- [Stakeholder Feedback / Opposition](#stakeholder-feedback--opposition)
+- [References & acknowledgements](#references--acknowledgements)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Introduction
+
+This is a proposal to support
+[Storage Foundation API's](https://github.com/WICG/storage-foundation-api-explainer)
```
Comment (rstz) [inline AccessHandle.md:36]: Consider reducing the focus on Storage Foundation API to improve readability

File: AccessHandle.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+applications through a very fast and generic storage backend. It especially
+focuses on ported WebAssembly programs for which existing storage web APIs
+don’t match the expectations of native code. There are a few design choices
+that primarily contribute to Storage Foundation’s performance:
+
+*   Developers get control of when data is flushed to disk, and reads can be
+    executed consistently without flushing. This allows programs to only schedule
+    time consuming flushes when they are required for persistency, and not as a
+    precondition to operate on recently written data.
+*   Its design ensures that implementations can access data from “the renderer
```
Comment (rstz) [inline AccessHandle.md:50]: I don't see how this is relevant for the Explainer, since the multi-process architecture is Chrome-specific and the quota system might not be part of the spec at all.

File: AccessHandle.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    process” (i.e. without needing to rely on a component that coordinates
+    between execution contexts) to avoid costly inter-process communication
+    (IPCs). Examples of this approach include the allocation-based quota system
+    and only allowing one open file handle per file.
+*   We avoid extra memory allocations by taking a buffer when reading.
+
+In order to bring this use cases to OPFS, we propose adding a method to
+*FileSystemFileHandle* that returns an *AccessHandle*, a new interface.
+*AccessHandles* provide the right kind of access and locking semantics to achieve
```
Comment (rstz) [inline AccessHandle.md:59]: What do you mean by "right kind of access"?

File: AccessHandle.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+[Origin Private File System](https://wicg.github.io/file-system-access/#sandboxed-filesystem),
+and doesn't currently consider changes to the rest of File System Access API or
+how files in the host machine are accessed.
+
+## Proposed API
+
+### New data access surface
+
+```javascript
+//In all contexts
```
Comment (rstz) [inline AccessHandle.md:104]: Consider adding whitspace after //

File: AccessHandle.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+const handle = await file.createAccessHandle();
+await handle.writable.getWriter().write(buffer);
```
Comment (rstz) [inline AccessHandle.md:106]: The "writable" here is a bit confusing, since it is not the writable that's created by "createWritable()", especially since these other writables are mentioned in Line 159. If you cannot rename them, maybe

File: AccessHandle.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+It  would return an *AccessHandle* that contains a
+[duplex stream](https://streams.spec.whatwg.org/#other-specs-duplex) and
+auxilliary methods. The readable/writable pair in the duplex stream would
+communicate with the same backing file, allowing the user to read unflushed
+contents. Another new method, *createSyncAccessHandle()*, would be only exposed
+on Worker threads. This method would offer a more buffer based surface for read
+and write. An IDL description of the new interface can be found in the
+[Appendix](#appendix).
+
+The reason for offering a Worker-only synchronous interface, is that consuming
```
Comment (rstz) [inline AccessHandle.md:127]: I think the comma is not needed?

File: AccessHandle.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+asynchronous APIs from Wasm has severe performance implications (more details
+[here](https://docs.google.com/document/d/1lsQhTsfcVIeOW80dr467Auud_VCeAUv2ZOkC63oSyKo)).
+We've opted for slightly different interfaces between the async and sync
+versions, because there currently isn’t support for synchronous streams. The
+amount of effort that would be required to spec them in a way that makes them
+compatible with asynchronous streams would be prohibitively high, and likely
+not worth it to support a single API.  Furthermore, we hope to discourage the
+use of the sync surface unless a developer is dealing with legacy Wasm ports,
+and therefore there is not much benefit to adding them.
```
Comment (rstz) [inline AccessHandle.md:136]: Could you rephrase that? It sounds like there is not much benefit in the sync interface at all

File: AccessHandle.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+This proposal assumes that
+[seekable streams](https://github.com/whatwg/streams/issues/1128) will be
+available. If this doesn’t happen, We can emulate the seeking behavior by
```
Comment (rstz) [inline AccessHandle.md:140]: We -> we

File: AccessHandle.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+} catch(e) {
+  //This catch will always be executed, since there is an open access handle
+}
+await handle1.close();
+//Now a new access handle may be created
+```
+
+In order to avoid multiple contexts modifying a file at the same time, locking
+semantics would be added to the new surface. At any given time, and across
+execution contexts, there would only be either a single *AccessHandle* per
```
Comment (rstz) [inline AccessHandle.md:158]: or SyncAccessHandle?

File: AccessHandle.md (hunk 1 above)
Comment (fivedots) [reply AccessHandle.md]: I've rewritten a big part of the doc, could you PTAL?

File: AccessHandle.md (hunk 2 above)
Comment (fivedots) [reply AccessHandle.md]: Removed.

File: AccessHandle.md (hunk 3 above)
Comment (fivedots) [reply AccessHandle.md]: Removed.

File: AccessHandle.md (hunk 4 above)
Comment (fivedots) [reply AccessHandle.md]: Done.

File: AccessHandle.md (hunk 5 above)
Comment (fivedots) [reply AccessHandle.md:86]: Sadly that name is part of the definition of a duplex stream, so there is benefit to keeping it. I think there's a part of yor comment missing, what is your suggestion?

File: AccessHandle.md (hunk 6 above)
Comment (fivedots) [reply AccessHandle.md:102]: Removed.

File: AccessHandle.md (hunk 7 above)
Comment (fivedots) [reply AccessHandle.md]: Done.

File: AccessHandle.md (hunk 8 above)
Comment (fivedots) [reply AccessHandle.md]: Done.

File: AccessHandle.md (hunk 9 above)
Comment (fivedots) [reply AccessHandle.md]: Done.

**Review by rstz — APPROVED** [review]
LGTM modulo comments

File: AccessHandle.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+[Exposed=DedicatedWorker]
+interface FileSystemSyncAccessHandle {
+  unsigned long long read([AllowShared] BufferSource buffer,
+                             FilesystemReadWriteOptions options);
+  unsigned long long write([AllowShared] BufferSource buffer,
+                              FilesystemReadWriteOptions options);
+
+  Promise<undefined> truncate([EnforceRange] unsigned long long size);
+  Promise<unsigned long long> getSize();
+  Promise<undefined> flush();
```
Comment (mkruisselbrink) [inline AccessHandle.md:193]: iirc, in the NativeIO/Storage Foundation API these were all synchronous as well. Could you elaborate on what you learned that made you change the API shape in this newer proposal?

File: AccessHandle.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    relying on the new surface's performance and direct buffered access to
+    offload sound segments to disk instead of holding them in memory.
+*   Provide a fast and persistent [Emscripten](https://emscripten.org/)
+    filesystem to act as generic and easily accessible storage for Wasm.
+
+## Non-goals
+
+This proposal is focused only on additions to the [Origin Private File
+System](https://wicg.github.io/file-system-access/#sandboxed-filesystem), and
+doesn't currently consider changes to the rest of File System Access API or how
```
Comment (mkruisselbrink) [inline AccessHandle.md:54]: While I know you're focusing on the origin private use cases specifically for this proposal, I think it would be worthwhile to explore what parts of this API could be exposed on arbitrary file handles as well. I believe there are a number of use cases that could be addressed by adding a couple of options to the createAccessHandle methods to slightly tweak its behavior.

Straw-man proposal `createAccessHandle({inPlace: true, locking: true})` would match what you've described here, and be only available for origin private files. 

On the other hand inPlace: false could then work with the same API surface for arbitrary file handles. While that would of course duplicate some API surface from the existing createWritable method, being able to read what has been written before it is flushed/comitted seems useful, and if this new API will be built on top of some better seekable streams primitive that might be a benefit as well.

Not sure if a non-locking version for the origin private file system makes sense, or is desirable, but for arbitrary files I think it would be useful to be able to create both locking and non-locking access handles.

While all that can probably be added after the fact, we can't really change the defaults for such options easily, so while it might perhaps make sense to initially only support a "inPlace: true, locking: true" mode (and throw/reject for any other options), I don't think that should be the default flags if we actually want to pursue a more unified API.

File: index.bs [hunk 12]
```diff
@@ -473,7 +473,7 @@ Advisement: In the Origin Trial as available in Chrome 82, createWritable replac
   :: Returns a {{FileSystemWritableFileStream}} that can be used to write to the file. Any changes made through
      |stream| won't be reflected in the file represented by |fileHandle| until the stream has been closed.
      User agents try to ensure that no partial writes happen, i.e. the file represented by
-     |fileHandle| will either contains its old contents or it will contain whatever data was written
+     |fileHandle| will either contain its old contents or it will contain whatever data was written
```
[the remark this answers is not in the document]
Comment (fivedots) [reply index.bs:476]: Removed the extraneous commit

File: AccessHandle.md (hunk 11 above)
Comment (fivedots) [reply AccessHandle.md:86]: That's a good point, I agree.

I've added a "mode" parameter to createAccessHandle, and an open question describing why it's there and what other values it could have. Please take a look and let me know if this covers what you meant!
